File: src/main/java/sample/design_pattern/State.java
/*
 * Copyright 2013 SatoShun
 *
 * State Pattern
 */

package sample.design_pattern;


interface IWeather {
    void compute();
}


class Morning implements IWeather {
    public void compute() {
        System.out.println("Good Morning");
    }
}


class Evening implements IWeather {
    public void compute() {
        System.out.println("Good Evening");
    }
}


class Night implements IWeather {
    public void compute() {
        System.out.println("Good Night");
    }
}


class State {
    public static void main(String[] args) {
        say(new Morning());
        say(new Night());
        say(new Evening());
    }

    private static void say(IWeather weather) {
        weather.compute();
    }
}

File: src/main/scala/sample/design_pattern/State.scala
/*
 * Copyright 2013 SatoShun
 *
 * State Pattern
 */

package sample.design_pattern;

trait IWeather {
    def compute(): Unit;
}


class Morning extends IWeather {
    def compute() = println("Good Morning")
}


class Evening extends IWeather {
    def compute() = println("Good Evening")
}


class Night extends IWeather {
    def compute() = println("Good Night")
}


object State {
    def main(args: Array[String]) = {
        say(new Morning())
        say(new Evening())
        say(new Night())
    }

    private def say(weather: IWeather) = weather.compute()
}
